### PR TITLE
Se2 y22 g21 101 add authentication system for users

### PR DESCRIPTION
**Pull request by Joja81**

### CODE REVIEW

File: app/functions/authentication.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+from pydoc import describe
+
+from sqlalchemy import desc
+from app.functions.email import validate_email
+from app.functions.error import AccessError, InputError
+from app.models import Session, User, db
+
+import jwt
+
+SECRET = os.environ.get('SECRET')
```
Comment (iamtayleungg) [inline app/functions/authentication.py:12]: Would be cute if we used something other than 'secret' :)

File: app/models.py [hunk 2]
```diff
@@ -20,6 +21,16 @@ class Call(db.Model):
     timeCalled = Column(DateTime(timezone=True), server_default=func.now())
 
     userAuth = Column(Boolean, nullable = False)
+
+class Session(db.Model):
```
Comment (iamtayleungg) [inline app/models.py:25]: Do I need to add this to ER diagram?

File: app/functions/authentication.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+        raise InputError(description="Session does not exist")
+    
+    return session.userId
+
+def hash(input):
+    '''
+    Generates hash using sha256 and returns as string
+    '''
+
+    return hashlib.sha256(input.encode()).hexdigest()
```
Comment (iamtayleungg) [inline app/functions/authentication.py:161]: The rest of this lgtm

File: app/models.py [hunk 4]
```diff
@@ -28,6 +39,10 @@ class User(db.Model):
     
     email = Column(String(320), nullable = False, unique = True)
     
-    token = Column(String(100), nullable = False, unique = True)
+    username = Column(String(100), nullable = False, unique = True)
```
Comment (iamtayleungg) [inline app/models.py:42]: Any reason why we're making the char limit 100 for all of them?

**Review by iamtayleungg — APPROVED** [review]
LGTM

**Review by iamtayleungg — CHANGES_REQUESTED** [review]
Wait please see comments first

File: app/functions/authentication.py (hunk 1 above)
Comment (Joja81) [reply app/functions/authentication.py:12]: Nah, secret is kinda standard name for the encryption key

File: app/models.py (hunk 2 above)
Comment (Joja81) [reply app/models.py:25]: Yep, sorry

File: app/models.py (hunk 4 above)
Comment (Joja81) [reply app/models.py:42]: Needed a max length for them and 100 seemed like a reasonable maximum.

Comment (Joja81) [reply app/models.py:42]: Matches the limits in interface spec

**Review by iamtayleungg — APPROVED** [review]
OK fr LGMT

**Pull request merged**